lib/openldap.c

```diff
@@ -231,40 +217,65 @@ static CURLcode oldap_map_error(int rc, CURLcode result)
   return result;
 }
 
-static CURLcode oldap_setup_connection(struct Curl_easy *data,
-                                       struct connectdata *conn)
+static CURLcode oldap_url_parse(struct Curl_easy *data, LDAPURLDesc **ludp)
 {
-  struct ldapconninfo *li;
-  LDAPURLDesc *lud;
-  int rc, proto;
-  CURLcode status;
+  CURLcode result = CURLE_OK;
+  int rc = LDAP_URL_ERR_BADURL;
+  static const char * const url_errs[] = {
+    "success",
+    "out of memory",
+    "bad parameter",
+    "unrecognized scheme",
+    "unbalanced delimiter",
+    "bad URL",
+    "bad host or port",
+    "bad or missing attributes",
+    "bad or missing scope",
+    "bad or missing filter",
+    "bad or missing extensions"
+  };
 
-  rc = ldap_url_parse(data->state.url, &lud);
+  *ludp = NULL;
+  if(!data->state.up.user && !data->state.up.password &&
+     !data->state.up.options)
+    rc = ldap_url_parse(data->state.url, ludp);
   if(rc != LDAP_URL_SUCCESS) {
     const char *msg = "url parsing problem";
-    status = CURLE_URL_MALFORMAT;
-    if(rc > LDAP_URL_SUCCESS && rc <= LDAP_URL_ERR_BADEXTS) {
-      if(rc == LDAP_URL_ERR_MEM)
-        status = CURLE_OUT_OF_MEMORY;
+
+    result = rc == LDAP_URL_ERR_MEM? CURLE_OUT_OF_MEMORY: CURLE_URL_MALFORMAT;
+    if(rc > LDAP_URL_SUCCESS && rc <= LDAP_URL_ERR_BADEXTS)
```

Comment:
I know this is better than before, but shouldn't this condition rather use the size of the string array instead so that we can be really **sure** that the following lookup never reads out of boundary by its own merit instead? That way we can verify that the code is fine without having to lookup what the contents of  `LDAP_URL_ERR_BADEXTS` is and has been been.

Reply:
Done. This is far better I think. Thanks for the suggestion.